### ISSUE TITLE
Make sure replay id is saved after processing callback

### ISSUE THREAD

**Issue by taman9333** [restforce/restforce#861]
I am willing to use a custom relay mechanism to ensure that no messages are dropped while my application is deploying. Therefore, I am utilizing something like the SimpleReplayHandler mentioned in the documentation. However, I have a question: if the subscription logic is as follows
```ruby
EM.run {
  # Subscribe to the PushTopic and use the custom replay handler to store any
  # received replay ID.
  client.subscription '/topic/AllAccounts', replay: SimpleReplayHandler.new do |message|
    puts message.inspect
  end
}
```
How I can make sure that the following method
```ruby
def []=(channel, replay_id)
    puts "[#{channel}] Writing replay ID: #{replay_id}"

    @last_modified = Time.now
    @channels[channel] = replay_id
  end
 ```
 will be executed after `puts message.inspect`, as if the `[]=(channel, replay_id)` method will be executed before the callback being executed then there might be an issue here when `[]=(channel, replay_id)` is executed then my application is deploying so it will restarts before `puts message.inspect` is executed

**Comment by taman9333** [comment]
@timrogers @panozzaj

**Comment by taman9333** [comment]
Would that work
```ruby
EM.run {
  replay_handler = SimpleReplayHandler.new

  # Subscribe to the PushTopic and use the custom replay handler to store any
  # received replay ID.
  client.subscription '/topic/AllAccounts', replay: replay_handler do |message|
    # Print the message
    puts message.inspect

    # Save the replay ID after printing the message
    replay_id = message['event']['replayId']
    channel = '/topic/AllAccounts'
    replay_handler[channel] = replay_id
  end
}
```

**Comment by panozzaj** [comment]
Hi Taman, thanks for opening a new issue. I have to admit that I have not worked with this repo for some time and am not a maintainer, but may be able to help.

You can ask Salesforce to replay events with the Rest force library. I think you could store what ids you have processed or what the last id was, and then when you start your app to request events since that id.

Please see https://github.com/restforce/restforce?tab=readme-ov-file#replaying-events and 
https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/using_streaming_api_durability.htm

If you need to be totally sure that you are not dropping messages, you may want to explore having a long-running standalone application that just listens to events and puts them on a message queue. Then it would not be affected by restarting.  But I suspect using the replay would help solve your problem.

**Comment by taman9333** [comment]
Hello panozzaj, first of all, I much appreciate your reply.

I have been running the logic for push topics in a standalone app. However, this app will be redeployed with new deployments we make, which might not be efficient, as we rarely add new changes to this standalone app.

Although I didn't have much context on the internals of the restforce gem itself, yesterday I was checking if the code mentioned in [the documentation](https://github.com/restforce/restforce?tab=readme-ov-file#replaying-events) would ensure that the logic to save the replay ID would run only after processing the message itself (the callback being executed).

I found that it's actually doing that, since in the incoming logic, [the callback will be executed first](https://github.com/restforce/restforce/blob/main/lib/restforce/concerns/streaming.rb#L66), and then [the logic to save the replay ID would run](https://github.com/restforce/restforce/blob/main/lib/restforce/concerns/streaming.rb#L73).

Once again, thank you very much for your comment. 🙏

**Comment by panozzaj** [comment]
OK, thanks, happy to try to help!

After re-reading the thread, I'm wondering if you still had any questions outstanding. I agree with your analysis that the callback is executed first, and then the id is saved. This seems good as long as your processing code is resilient to the callback potentially getting executed twice. For example, if you had the following happen:

1. receive push topic id 123
2. execute callback for id 123
3. record id 123
4. receive push topic id 124
5. execute callback for id 124
6. but let's say the app is restarted here, so 124 is only partially processed or at least the id is not saved
7. upon restart, ask Salesforce to replay all topics since 123
8. then it would send 124 and on, so potentially reprocessing 124 or parts of 124

At least that's my current best understanding.

**Comment by taman9333** [comment]
Yep, the edge case you mentioned is correct, and my code is okay to execute the callback twice. Again, thanks a lot for your help.
